mlir/lib/Conversion/EmulateFp8ExtTrunc/EmulateFp8ExtTrunc.cpp

```diff
@@ -365,14 +384,264 @@ static FlatSymbolRefAttr makeFp8TruncFunction(Location loc, FloatType outType,
+  b.create<cf::BranchOp>(bb9, ValueRange{i32Const(-126), add46, and22});
+
+  b.setInsertionPointToStart(bb8);
+  Value add47 = b.create<AddIOp>(shl42, i32Const(2));
+  Value add48 = b.create<AddIOp>(and24, i32Const(-127));
+  Value cmp49 = b.create<CmpIOp>(CmpIPredicate::sgt, add48, add47);
+  Value sub50 = b.create<SubIOp>(add47, add48);
+  Value or51  = b.create<OrIOp>(and22, i32Const(0x800000));
+  Value select52 = b.create<SelectOp>(cmp49, i32Const(0), sub50);
+  b.create<cf::BranchOp>(bb9, ValueRange{add48, select52, or51});
+
+  Block *bb10 = func.addBlock();
+  Block *bb11 = func.addBlock();
+  Block *bb12 = func.addBlock();
+  bb12->addArgument(i32, loc);
+  b.setInsertionPointToStart(bb9);
+  Value bb9arg0 = bb9->getArgument(0);
+  Value bb9arg1 = bb9->getArgument(1);
+  Value bb9arg2 = bb9->getArgument(2);
+  Value sub56 = b.create<SubIOp>(i32Const(23), mWidth);
+  Value add57 = b.create<AddIOp>(bb9arg1, sub56);
+  Value min58 = b.create<MinUIOp>(add57, i32Const(31));
+  Value shl59 = b.create<ShLIOp>(i32Const(-1), min58);
+  Value xor60 = b.create<XOrIOp>(shl59, i32Const(-1));
+  Value and61 = b.create<AndIOp>(bb9arg2, xor60);
+  Value add62 = b.create<AddIOp>(add57, i32Const(-1));
+  Value min63 = b.create<MinUIOp>(add62, i32Const(31));
+  Value shl64 = b.create<ShLIOp>(i32Const(1), min63);
+  Value cmp65 = b.create<CmpIOp>(CmpIPredicate::eq, and61, shl64);
+  Value cmp66 = b.create<CmpIOp>(CmpIPredicate::sgt, bb9arg1, i32Const(0));
+  b.create<cf::CondBranchOp>(cmp66, bb10, bb11);
+
+  b.setInsertionPointToStart(bb10);
+  Value min67 = b.create<MinUIOp>(bb9arg1, i32Const(31));
+  Value shr68 = b.create<ShRUIOp>(bb9arg2, min67);
+  b.create<cf::BranchOp>(bb12, ValueRange{shr68});
+
+  b.setInsertionPointToStart(bb11);
+  Value cmp69 = b.create<CmpIOp>(CmpIPredicate::eq, bb9arg1, i32Const(-1));
+  Value zext70 = b.create<ExtUIOp>(i32, cmp69);
+  Value shl71 = b.create<ShLIOp>(bb9arg2, zext70);
+  b.create<cf::BranchOp>(bb12, ValueRange{shl71});
+
+  Block *bb13 = func.addBlock();
+  Block *bb14 = func.addBlock();
+  Block *bb15 = func.addBlock();
+  bb15->addArgument(i32, loc);
+  bb15->addArgument(i32, loc);
+  b.setInsertionPointToStart(bb12);
+  Value bb12arg0 = bb12->getArgument(0);
+  Value shr73 = b.create<ShRUIOp>(bb12arg0, i32Const(23));
+  Value or74  = b.create<OrIOp>(shr73, i32Const(-2));
+  Value sub75 = b.create<SubIOp>(bb9arg0, shl42);
+  Value add76 = b.create<AddIOp>(sub75, bb9arg1);
+  Value add77 = b.create<AddIOp>(add76, or74);
+  Value shl78 = b.create<ShLIOp>(i32Const(1), sub56);
+  Value add79 = b.create<AddIOp>(shl78, i32Const(-1));
+  Value and80 = b.create<AndIOp>(bb12arg0, shl78);
+  Value cmp81 = b.create<CmpIOp>(CmpIPredicate::eq, and80, i32Const(0));
+  Value select82 = b.create<SelectOp>(cmp65, cmp81, i1Const(false));
+  Value sext83 = b.create<ExtSIOp>(i32, select82);
+  Value add84 = b.create<AddIOp>(bb12arg0, sext83);
+  Value stoch = i1Const(false);         // Defaulted arguments.
```

Comment:
The C++ cast_to_f8<> that I used had stoch (for stochastic rounding) and rng (value used for that) as parameters with default values.  OCP spec mandates round-to-nearest-even but allows other rounding modes.  Hard-code the defaults here;  if we want to extend it, this is where to hook in the control.